### DISCUSSION TITLE
Code completion not available in Swift Embedded project

### DISCUSSION THREAD

**Discussion by DevSrSouza** [swiftlang/vscode-swift#968] [Q&A]
Hey folks! This WWDC they have shared the [Swift Embedded](https://github.com/swiftlang/swift/blob/main/docs/EmbeddedSwift/UserManual.md) support with cool samples using Matter. Here is the video https://developer.apple.com/videos/play/wwdc2024/10197/ .

In the video, they use NeoVim and it does have code completion suggestions, I have being trying to make it work with the VSCode Swift, but without success. I'm using latest Swift version (6.0 nightly) and setup in the VSC plugin as the toolchain but it did not improve.

Here is the video sample project if you may want to test
https://github.com/apple/swift-matter-examples/tree/main/led-blink

It is required to install esp-idf and esp-matter, I had to install Python 3.9.2 via ASDF, the current MacOS version does not work/compile esp-matter.

**Comment by waqi786** [comment]
Hi 
Thank you for sharing your experience. It sounds like you've encountered some challenges with setting up Swift Embedded support in VSCode, similar to my own experience.

I've been exploring Swift Embedded with Matter as well and faced issues with code completion in VSCode. One approach that has helped me is ensuring the Swift Language Server (SourceKit-LSP) is correctly configured:

1. Install SourceKit-LSP: You can install it via Swift's toolchain.
         swift build -Xswiftc -swift-version -Xswiftc 6.0
2. Configure VSCode: Ensure your settings.json in VSCode includes the path to the Swift toolchain.
         {
  "swift.path.swift_driver_bin": "/path/to/swift/bin",
  "swift.path.sourcekit-lsp": "/path/to/sourcekit-lsp"
}

3. Check Extensions: Make sure you have the latest Swift extension installed from the VSCode marketplace.

4. Reopen Project: Sometimes, simply reopening the project in VSCode can refresh the settings and improve code completion.

Additionally, verify if there are any specific configurations needed for the embedded environment you're using. The sample project you mentioned seems complex, so ensuring all dependencies and environment variables are correctly set is crucial.

Here’s the link to the SourceKit-LSP repository for more detailed setup instructions: https://github.com/apple/sourcekit-lsp.

I hope this information helps! If you discover any other solutions or need further assistance, please feel free to share.

**Reply by adam-fowler** [reply]
I'm a little confused by this response. The swift extension (published by SSWG) does not have settings "swift.path.swift_driver_bin" or "swift.path.sourcekit-lsp". If you are using macOS to build these examples you probably want to install the latest Xcode beta to get Swift 6. Then in VSCode you need to choose the toolchain that comes with the Xcode beta. Run command `Swift: Select Toolchain...` to do this.

You should not need to build SourceKit-LSP the version that comes with the Xcode beta toolchain should suffice.

I personally don't have any experience of the embedded support. From the examples it looks like it is using CMake files to build. SourceKit-LSP should pick up the details it needs from the CMake compile_commands.json file. Currently SourceKit-LSP requires a built project for source completion etc to work. So build your project, symbol completion should work after that eventually, you can maybe be certain of that if you restart the extension using the command `Developer: Reload Window`

**Reply by DevSrSouza** [reply]
The `compile_commands.json` is available inside `build/`, I even did a symlink to the workspace folder but it does not work as well.

This is mine settings.json vsc swift settings, as you can see, I have already select the proper Swift toolchain, I don't have Xcode Beta installed, but at my Xcode and zshrc they are both configured with the same toolchain bellow.
```json
"swift.path": "/Users/homemac/Library/Developer/Toolchains/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-07-10-a.xctoolchain/usr/bin",
    "swift.swiftEnvironmentVariables": {
        "DEVELOPER_DIR": "/Applications/Xcode.app"
    }
```

@adam-fowler there is any way to see logs or improve logs from the vsc swift plugin and sourcekit-lsp that could help to figure out?

**Reply by DevSrSouza** [reply]
<img width="262" alt="image" src="https://github.com/user-attachments/assets/a9c90058-2c7e-4402-884b-ff7506bed62c">
<img width="543" alt="image" src="https://github.com/user-attachments/assets/03c14309-a469-4c6e-913f-154a7a84aa12">

**Reply by DevSrSouza** [reply]
Here you can see the `compile_commands.json` contains both Swift files + a bunch more from esp and matter sdk. If you want the raw file, let me know.
<img width="1479" alt="image" src="https://github.com/user-attachments/assets/25ee8869-89cb-40dc-b8e4-a805e8538407">

**Reply by plemarquand** [reply]
@DevSrSouza If you could run `Swift: Capture VS Code Swift Diagnostic Bundle` from the VS Code command palette and zip + attach it here that might help diagnose the issue.

**Reply by DevSrSouza** [reply]
Sure, here is the [files](https://github.com/user-attachments/files/16238613/Archive.zip)

**Reply by ahoppen** [reply]
It looks like SourceKit-LSP does not pick up your `compile_commands.json`.

Now the question is why, which is hard to tell because the log output redacts paths and file contents by default. Are you able to enable extended logging as described [here](https://github.com/swiftlang/sourcekit-lsp/blob/main/Documentation/Diagnose%20Bundle.md#enable-extended-logging), reproduce the issue and then capture a new diagnostic bundle? If you could also zip up your source folder and attach it, that might help as well.

**Comment by BrianHenryIE** [comment]
I tried to get this working today, unsuccessfully. I've been working with [apple/swift-embedded-examples:esp32-led-blink-sdk](https://github.com/apple/swift-embedded-examples/tree/main/esp32-led-blink-sdk).

I found adding `.vscode/c_cpp_properties.json`

```json
{
    "configurations": [
        {
            "name": "ESP-IDF",
            "compilerPath": "/Users/brianhenry/.espressif/tools/riscv32-esp-elf/esp-12.2.0_20230208/riscv32-esp-elf/bin/riscv32-esp-elf-gcc",
            "compileCommands": "${config:idf.buildPath}/compile_commands.json",
            "includePath": [
                "${config:idf.espIdfPath}/components/**",
                "${workspaceFolder}/**"
            ],
            "browse": {
                "path": [
                    "${config:idf.espIdfPath}/components",
                    "${workspaceFolder}"
                ],
                "limitSymbolsToIncludedHeaders": true
            }
        }
    ],
    "version": 4
}
```

got the ESP library code completion working in my `Dummy.c` file, e.g. I could navigate to the function source in:

```
#include "esp_log.h"

void never_called(void)
{
    ESP_LOGI("example", "log");
}
```

but `Main.swift` and `Led.swift` still had no "Go to definition" etc working, either between the two `.swift` files, or even to the ESP functions that otherwise were now working in the `.c` file.

I learned there is a [`.sourcekit-lsp/config.json` file](https://github.com/swiftlang/sourcekit-lsp/blob/main/Documentation/Configuration%20File.md) but it didn't have anything useful here.

My `build/compile_commands.json` and `build/bootloader/compile_commands.json` do _not_ have any mention of any `.swift` files. I tried manually editing one to reference `build/esp-idf/main/_swiftcode.o` but I didn't get it working (although at least I got red squiggles).

I could be wrong, but I think it's a two-step compile process where the output of the original step is overwritten by the second.

I spent a while trying to create a SPM package with the ESP [library as a submodule](https://www.swift.org/documentation/articles/wrapping-c-cpp-library-in-swift.html). 

```
[submodule "esp-idf"]
	path = Sources/CESPIDF/esp-idf
	url = /Users/brianhenry/esp/v5.1-rc2/esp-idf/
```

<details>

<summary>Package.json</summary>

```swift
// swift-tools-version: 5.9
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "swift-esp32-wrapper",
    products: [
        // Products define the executables and libraries a package produces, making them visible to other packages.
        .library(
            name: "swift-esp32-wrapper",
            targets: ["swift-esp32-wrapper"]),
    ],
    targets: [
        // Targets are the basic building blocks of a package, defining a module or a test suite.
        // Targets can depend on other targets in this package and products from dependencies.
        .target(
            name: "swift-esp32-wrapper"
            
            ),
         .executableTarget(
            name: "CESPIDF",
            dependencies: [],
            exclude: [
                // Relative paths under 'CMyLib' of the files
                // and/or directories to exclude. For example:
                // "./my-lib/src/CMakeLists.txt",
                // "./my-lib/tests",
            ],
            sources: [
                // "esp-idf/components/app_trace",
                // "esp-idf/components/app_update",
                // "esp-idf/components/bootloader",
                // "esp-idf/components/bootloader_support",
                // "esp-idf/components/bt",
                // "esp-idf/components/cmock",
                // "esp-idf/components/console",
                // "esp-idf/components/cxx",
                // "esp-idf/components/driver",
                // "esp-idf/components/efuse",
                // "esp-idf/components/esp-tls",
                // "esp-idf/components/esp_adc",
                // "esp-idf/components/esp_app_format",
                // "esp-idf/components/esp_coex",
                // "esp-idf/components/esp_common",
                // "esp-idf/components/esp_eth",
                // "esp-idf/components/esp_event",
                // "esp-idf/components/esp_gdbstub",
                // "esp-idf/components/esp_hid",
                // "esp-idf/components/esp_http_client",
                // "esp-idf/components/esp_http_server",
                // "esp-idf/components/esp_https_ota",
                // "esp-idf/components/esp_https_server",
                // "esp-idf/components/esp_hw_support",
                // "esp-idf/components/esp_lcd",
                // "esp-idf/components/esp_local_ctrl",
                // "esp-idf/components/esp_mm",
                // "esp-idf/components/esp_netif",
                // "esp-idf/components/esp_netif_stack",
                // "esp-idf/components/esp_partition",
                // "esp-idf/components/esp_phy",
                // "esp-idf/components/esp_pm",
                // "esp-idf/components/esp_psram",
                // "esp-idf/components/esp_ringbuf",
                // "esp-idf/components/esp_rom",
                // "esp-idf/components/esp_system",
                // "esp-idf/components/esp_timer",
                // "esp-idf/components/esp_wifi",
                // "esp-idf/components/espcoredump",
                // "esp-idf/components/esptool_py",
                // "esp-idf/components/fatfs",
                // "esp-idf/components/freertos",
                // "esp-idf/components/hal",
                // "esp-idf/components/heap",
                // "esp-idf/components/http_parser",
                // "esp-idf/components/idf_test",
                // "esp-idf/components/ieee802154",
                // "esp-idf/components/json",
                // "esp-idf/components/log",
                // "esp-idf/components/lwip",
                // "esp-idf/components/mbedtls",
                // "esp-idf/components/mqtt",
                // "esp-idf/components/newlib",
                // "esp-idf/components/nvs_flash",
                // "esp-idf/components/openthread",
                // "esp-idf/components/partition_table",
                // "esp-idf/components/protobuf-c",
                // "esp-idf/components/protocomm",
                // "esp-idf/components/pthread",
                // "esp-idf/components/riscv",
                // "esp-idf/components/sdmmc",
                // "esp-idf/components/soc",
                // "esp-idf/components/spi_flash",
                // "esp-idf/components/spiffs",
                // "esp-idf/components/tcp_transport",
                // "esp-idf/components/ulp",
                // "esp-idf/components/unity",
                // "esp-idf/components/usb",
                // "esp-idf/components/vfs",
                // "esp-idf/components/wear_levelling",
                // "esp-idf/components/wifi_provisioning",
                // "esp-idf/components/wpa_supplicant"
            ],
            cSettings: [
                .headerSearchPath(
                    "esp-idf/components/freertos/FreeRTOS-Kernel/include"
                    ),
                .unsafeFlags(["-w"]), // Disable all warnings
            ],
            swiftSettings: [
                // To mute warning about "... was identified as an executable target given the presence of a 'main' file."
                .unsafeFlags(["-parse-as-library"])
            ]
        ),
    ]
)
```
</details>

What I think I'm working towards is a point where Swift Package Manager thinks it knows enough to build the project, but ESP-IDF extension _actually_ builds the artifact that is deployed.

AIUI there are two ways to get Swift-LSP to understand a project – a CMake outputted `compile_commands.json`, or the build output of SPM (I don't know what specific file is used). I think the effort it would take me to understand and manipulate the output of the CMake compile should just as easily be used to get the SPM approach to work.

This looks promising for the first approach – [jerrymarino/SwiftCompilationDatabase](https://github.com/jerrymarino/SwiftCompilationDatabase) – but I've been working on this all day (Saturday) and I'm getting tired!

> It is required to install esp-idf and esp-matter, I had to install Python 3.9.2 via ASDF

I installed the ESP VS Code extension and it installed the ESP library which contained its own copy of Python which worked for me.

It's important to `source ~/esp/v5.1-rc2/esp-idf/export.sh` when running anything in Terminal.

And regularly check `which swift`. I'm sure it was using the nightly version in VS Code's Terminal but then at some point I noticed it was using `/usr/bin/swift`.

I had to add `set(SWIFTC "/Users/brianhenry/Library/Application Support/swiftly/bin/swiftc")` in `main/CMakeLists.txt` to build the project in VS Code – if you're building in a Terminal that has the correct Swift toolchain set, it inherits it, but the ESP extension always uses `/usr/bin/swift` for me.

@plemarquand , has anyone had code completion working for embedded yet? I don't think it's a "problem" with DevSrSouza's setup, I think it's a new configuration we need a solution for. That said, I did learn a bit from the files that were shared.

@adam-fowler I'm pretty sure the original answer was Chat GPT generated.

**Reply by ahoppen** [reply]
> My `build/compile_commands.json` and `build/bootloader/compile_commands.json` do _not_ have any mention of any `.swift` files

Which version of CMake are you running (`cmake --version`)? CMake only started generating compile commands for Swift files somewhat recently (I think 3.27 but it could also be 3.28 or 3.29). Updating CMake might fix that issue.

**Reply by etcwilde** [reply]
It's CMake 3.29, though the minimum required version for the project is set to [3.29](https://github.com/apple/swift-embedded-examples/blob/b9bc89abcb490cb801f8662d6df8f828aa6f43ad/esp32-led-blink-sdk/CMakeLists.txt#L1), so that should be setting everything you need.

**Comment by Joannis** [comment]
I've used Embedded Swift with VSCode, and documented my experience in this repo:
https://github.com/Joannis/pico-embedded-swift

Last time I ran the project has been two months ago, so some things might have changed a little bit. But I think that this should give some insight into what's needed to get up and running.

I've had to quite carefully craft my CMakeLists.txt. Most notably, I needed specific policies to use SourceKit-LSP and export the compile commands. Without those the IDE's functionally is reduced to just text highlighting.

The Compile Commands are essential, and if you experience any issues I recommend opening the file to see if the paths make sense. Some (olderr) builds of CMake or the Swift compiler don't work together well enough, and the paths become nonsensical and thus unsable for VSCode.

Once it's set up though, VSCode can follow code between Swift, C and C++ all the same.

**Comment by BrianHenryIE** [comment]
Thanks @Joannis. That's optimistic but I wasn't able to get your repo to build! I keep getting:

> ninja: error: 'CMakeFiles/SwiftLib.dir/SwiftLib.swift.obj', needed by 'SwiftLib', missing and no known rule to make it

although I think that should be built by the command

```
add_library(SwiftLib OBJECT
    SwiftLib.swift
    SwiftPicoPrint.swift
)
```

<details>

<summary>Full CMake build output</summary>

```
[main] Building folder: /Users/brianhenry/Sites/swift-esp32-led-blink/pico-embedded-swift/build 
[main] The folder containing the CMake cache is missing. The cache will be regenerated.
[main] Configuring project: pico-embedded-swift 
[proc] Executing command: /Users/brianhenry/.pico-sdk/cmake/v3.28.6/bin/cmake -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE "-DCMAKE_C_COMPILER:FILEPATH=/Users/brianhenry/Library/Application Support/swiftly/bin/clang" "-DCMAKE_CXX_COMPILER:FILEPATH=/Users/brianhenry/Library/Application Support/swiftly/bin/clang++" --no-warn-unused-cli -S/Users/brianhenry/Sites/swift-esp32-led-blink/pico-embedded-swift -B/Users/brianhenry/Sites/swift-esp32-led-blink/pico-embedded-swift/build -G Ninja
[cmake] --- START ---
[cmake] Not searching for unused variables given on the command line.
[cmake] Swift toolchain: ('org.swift.59202407251a') (using value from CMakeLists.txt as TOOLCHAINS is not set in environment)
[cmake] PICO_SDK_PATH is /Users/brianhenry/.pico-sdk/sdk/1.5.1
[cmake] Defaulting PICO_PLATFORM to rp2040 since not specified.
[cmake] Defaulting PICO platform compiler to pico_arm_gcc since not specified.
[cmake] PICO compiler is pico_arm_gcc
[cmake] -- The C compiler identification is GNU 13.2.1
[cmake] -- The CXX compiler identification is GNU 13.2.1
[cmake] -- The ASM compiler identification is GNU
[cmake] -- Found assembler: /Users/brianhenry/.pico-sdk/toolchain/13_2_Rel1/bin/arm-none-eabi-gcc
[cmake] -- The Swift compiler identification is Apple 6.0
[cmake] -- Detecting C compiler ABI info
[cmake] -- Detecting C compiler ABI info - done
[cmake] -- Check for working C compiler: /Users/brianhenry/.pico-sdk/toolchain/13_2_Rel1/bin/arm-none-eabi-gcc - skipped
[cmake] -- Detecting C compile features
[cmake] -- Detecting C compile features - done
[cmake] -- Detecting CXX compiler ABI info
[cmake] -- Detecting CXX compiler ABI info - done
[cmake] -- Check for working CXX compiler: /Users/brianhenry/.pico-sdk/toolchain/13_2_Rel1/bin/arm-none-eabi-gcc - skipped
[cmake] -- Detecting CXX compile features
[cmake] -- Detecting CXX compile features - done
[cmake] Build type is Debug
[cmake] Using regular optimized debug build (set PICO_DEOPTIMIZED_DEBUG=1 to de-optimize)
[cmake] Defaulting PICO target board to pico since not specified.
[cmake] Using board configuration from /Users/brianhenry/.pico-sdk/sdk/1.5.1/src/boards/include/boards/pico.h
[cmake] -- Found Python3: /opt/anaconda3/bin/python3.12 (found version "3.12.4") found components: Interpreter 
[cmake] TinyUSB available at /Users/brianhenry/.pico-sdk/sdk/1.5.1/lib/tinyusb/src/portable/raspberrypi/rp2040; enabling build support for USB.
[cmake] Compiling TinyUSB with CFG_TUSB_DEBUG=1
[cmake] BTstack available at /Users/brianhenry/.pico-sdk/sdk/1.5.1/lib/btstack
[cmake] cyw43-driver available at /Users/brianhenry/.pico-sdk/sdk/1.5.1/lib/cyw43-driver
[cmake] Pico W Bluetooth build support available.
[cmake] lwIP available at /Users/brianhenry/.pico-sdk/sdk/1.5.1/lib/lwip
[cmake] mbedtls available at /Users/brianhenry/.pico-sdk/sdk/1.5.1/lib/mbedtls
[cmake] -- Building Pico-W CYW43439 version
[cmake] -- Configuring done (2.4s)
[cmake] -- Generating done (0.1s)
[cmake] -- Build files have been written to: /Users/brianhenry/Sites/swift-esp32-led-blink/pico-embedded-swift/build
[build] Starting build
[proc] Executing command: /Users/brianhenry/.pico-sdk/cmake/v3.28.6/bin/cmake --build /Users/brianhenry/Sites/swift-esp32-led-blink/pico-embedded-swift/build --config Debug --target all --
[build] ninja: error: 'CMakeFiles/SwiftLib.dir/SwiftLib.swift.obj', needed by 'SwiftLib', missing and no known rule to make it
[proc] The command: /Users/brianhenry/.pico-sdk/cmake/v3.28.6/bin/cmake --build /Users/brianhenry/Sites/swift-esp32-led-blink/pico-embedded-swift/build --config Debug --target all -- exited with code: 1
[driver] Build completed: 00:00:00.021
[build] Build finished with exit code 1
```

</details>

I did make some minor changes to `CMakeLists.txt` – the Raspberry Pi Pico VS Code Extension added a block of code at the top. Then I had to define `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER` because when I ran CMake first VS Code asked me to select a compiler and I picked Swift's clang which does not recognise `-mcpu=cortex-m0plus`. Plus the obvious `swiftc` change.

<details>

<summary>CMakeLists.txt</summary>

```
# == DO NEVER EDIT THE NEXT LINES for Raspberry Pi Pico VS Code Extension to work ==
if(WIN32)
   set(USERHOME $ENV{USERPROFILE})
else()
    set(USERHOME $ENV{HOME})
endif()
set(PICO_SDK_PATH ${USERHOME}/.pico-sdk/sdk/1.5.1)
set(PICO_TOOLCHAIN_PATH ${USERHOME}/.pico-sdk/toolchain/13_2_Rel1)
if(WIN32)
    set(pico-sdk-tools_DIR ${USERHOME}/.pico-sdk/tools/1.5.1)
    include(${pico-sdk-tools_DIR}/pico-sdk-tools-config.cmake)
    include(${pico-sdk-tools_DIR}/pico-sdk-tools-config-version.cmake)
endif()
# ====================================================================================
message("--- START ---")

cmake_minimum_required(VERSION 3.28)

# Without these policies, Sourcekit-LSP won't work
if(POLICY CMP0126)
cmake_policy(SET CMP0126 NEW)
endif()
if(POLICY CMP0157)
cmake_policy(SET CMP0157 NEW)
endif()

# User configuration
# Change these values to match your config.
# 
# 1) Path to Raspberry Pi Pico C/C++ SDK
# Defined above by the vs-code extension.

# 2) Swift toolchain 
#    Change this to the current Swift nightly toolchain you installed.
#    The ID is in /Library/Developer/Toolchains/[TOOLCHAIN].xctoolchain/Info.plist
#    Alternatively, set the TOOLCHAINS environment variable.

# export TOOLCHAINS=$(plutil -extract CFBundleIdentifier raw /Users/brianhenry/Library/Developer/Toolchains/swift-latest.xctoolchain/Info.plist)

if (NOT (DEFINED ENV{TOOLCHAINS}))
    set(Swift_Toolchain "org.swift.59202407251a")
    message("Swift toolchain: ('${Swift_Toolchain}') (using value from CMakeLists.txt as TOOLCHAINS is not set in environment)")
else ()
    set(Swift_Toolchain "$ENV{TOOLCHAINS}")
    message("Swift toolchain: ('${Swift_Toolchain}') (using TOOLCHAINS value from environment)")
endif ()

# initialize the SDK based on PICO_SDK_PATH
# note: this must happen before project()
include(pico_sdk_import.cmake)

# Configure Swift. This must happen before `project()`. I don't know why.
# Use nightly Swift compiler, configured for Embedded Swift.
# Find path to swiftc and store it in swiftc_Path
execute_process(
    COMMAND xcrun --toolchain "${Swift_Toolchain}" --find swiftc
    OUTPUT_VARIABLE swiftc_Path
    OUTPUT_STRIP_TRAILING_WHITESPACE
)
set(CMAKE_Swift_COMPILER
    "${swiftc_Path}"
)
string(APPEND CMAKE_Swift_FLAGS
    # -wmo: Whole-module optimization is always required for Embedded Swift.
    # -Xfrontend -function-sections: enables dead stripping of unused runtime functions.
    " -target armv6m-none-none-eabi \
    -enable-experimental-feature Embedded \
    -enable-experimental-feature TypedThrows \
    -wmo \
    -Xfrontend -function-sections"
)
# Disable CMake’s automatic Swift compiler check. The compiler check always 
# fails for Embedded Swift because it tries to compile a Swift program that 
# includes `print()`, which isn't available in Embedded Swift.
set(CMAKE_Swift_COMPILER_FORCED TRUE)

set(SWIFTC ${USERHOME}/Application\ Support/swiftly/bin/swiftc)
set(CMAKE_C_COMPILER ${PICO_TOOLCHAIN_PATH}/bin/arm-none-eabi-gcc)
set(CMAKE_CXX_COMPILER ${PICO_TOOLCHAIN_PATH}/bin/arm-none-eabi-gcc)
# set(CMAKE_CXX_COMPILER "${PICO_TOOLCHAIN_PATH}/bin/arm-none-eabi-g++")

project(SwiftPico LANGUAGES C CXX ASM Swift)

# Initialize the Raspberry Pi Pico SDK
pico_sdk_init()

set(CMAKE_C_STANDARD 11)
set(CMAKE_CXX_STANDARD 17)

if (PICO_SDK_VERSION_STRING VERSION_LESS "1.3.0")
    message(FATAL_ERROR "Raspberry Pi Pico SDK version 1.3.0 (or later) required. Your version is ${PICO_SDK_VERSION_STRING}")
endif()

include_directories(${CMAKE_CURRENT_SOURCE_DIR}/picowi/lib)
add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/picowi/lib)

# https://cmake.org/cmake/help/latest/command/add_library.html#object-libraries
add_library(SwiftLib OBJECT
    SwiftLib.swift
    SwiftPicoPrint.swift
)

# set_target_properties(SwiftLib PROPERTIES Swift_COMPILATION_MODE "wholemodule")

add_dependencies(SwiftLib picowi)
add_dependencies(SwiftLib pico_stdlib)
add_dependencies(SwiftLib hardware_pio)
add_dependencies(SwiftLib hardware_dma)

add_executable(SwiftPico
    main.c
)

set_target_properties(SwiftPico PROPERTIES Swift_COMPILATION_MODE "wholemodule")

target_link_libraries(SwiftPico
    $<TARGET_OBJECTS:SwiftLib>
    pico_stdlib
    picowi hardware_pio hardware_dma
)

# Enable USB output, disable UART output
pico_enable_stdio_usb(SwiftPico 1)
pico_enable_stdio_uart(SwiftPico 0)

# create map/bin/hex file etc.
pico_add_extra_outputs(SwiftPico)
```

</details>

I'd appreciate any help. I've very little experience with CMake. 

Besides that, I tried [jerrymarino/SwiftCompilationDatabase](https://github.com/jerrymarino/SwiftCompilationDatabase/) but I wasn't able to capture the output of `swiftc` during the CMake step using [`swiftc ... -parseable-output 2>&1 > parseable-build.log`](https://github.com/jerrymarino/SwiftCompilationDatabase/blob/810f040e97ebe97210d64a644d8d386e35a6358d/Sources/SwiftCompilationDatabase/main.swift#L19).

I learned a lot from a reply by @etcwilde on [How to use SourceKit-LSP to develop the Swift compiler repo?](https://forums.swift.org/t/how-to-use-sourcekit-lsp-to-develop-the-swift-compiler-repo/66167/2) where I tried to use `ninja -t compdb > compile_commands.json` but ultimately it didn't work to get code-navigation working. I tried merging, overwriting, and placing `compile_commands.json` in the project root. And I was running Restart LSP Server between changes.

Another copy of the project where I was trying to restructure the project as a Swift Package Manager project does have code-navigation between Swift-to-Swift and C-to-C but not from Swift to C. As I tried adding the C sources to Swift Package Manager, the build time became extremely long and I still didn't get the code-navigation as I wanted.

@adam-fowler Am I correct to think vscode-swift should read _both_ SPM's output and CMake's `compile_commands.json` for LSP?

I even ordered a RPi Pico for this but Amazon  "[failed delivery attempt due to dog hazard](https://www.instagram.com/p/CzHJ4STxNm_/?img_index=3)"!

**Reply by adam-fowler** [reply]
The VSCode extension (or more specifically SourceKit-LSP) works with both SwiftPM and CMake projects. As I said earlier it uses the compile_commands.json to identify a CMake project.

**Reply by BrianHenryIE** [reply]
Thanks. I mean, does the presence of one supersede the existence of the other? Are both outputs used together or are they mutually exclusive?

**Reply by ahoppen** [reply]
SourceKit-LSP will pick one build system integration. It currently has the following precedence order: Build server, SwiftPM (existence of a Package.swift), compilation database (existence of `compile_commands.json`). If you want, you can override that behavior by setting the `defaultWorkspaceType` in your SourceKit-LSP [configuration file](https://github.com/swiftlang/sourcekit-lsp/blob/main/Documentation/Configuration%20File.md).

**Comment by adam-fowler** [comment]
They are mutually exclusive as far as I know. This would be a question for
SourceKit-LSP.

**Comment by BrianHenryIE** [comment]
Here's a promising project where the author says he has it working:
* https://github.com/anders0nmat/swift-embedded-pm
* https://forums.swift.org/t/swift-packages-for-esp-idf/73360

I spent yesterday, before I saw that post, trying to rewrite the CMakeLists so the Swift compilation is done by CMake rather than through an [`add_custom_command()`](https://github.com/apple/swift-embedded-examples/blob/17ab15e917b829498234a9899783b73ecf2c24a7/esp32-led-blink-sdk/main/CMakeLists.txt#L31-L48) step. 

There's a new reply on the forum post I linked above:
* https://forums.swift.org/t/how-to-use-sourcekit-lsp-to-develop-the-swift-compiler-repo/66167/5

Which itself links to a good thread on the topic:
* https://forums.swift.org/t/updates-for-swift-in-the-upcoming-cmake-3-29/70154

The other example embedded projects' CMakeLists files, both in [apple/swift-embedded-examples](https://github.com/apple/swift-embedded-examples) and [apple/swift-matter-examples](https://github.com/apple/swift-matter-examples), take slightly different approaches. I'd bet some might work with LSP out of the box. 

I'll take another look at this next weekend. I feel I'm almost there.

**Comment by BrianHenryIE** [comment]
I pulled the latest [apple/swift-embedded-examples](https://github.com/apple/swift-embedded-examples/) and when I built the [esp32-led-blink-sdk](https://github.com/apple/swift-embedded-examples/tree/main/esp32-led-blink-sdk) I immediately got code-completion and navigation, thanks @kubamracek 

And thanks everyone else for the help.

https://github.com/apple/swift-embedded-examples/commit/b9bc89abcb490cb801f8662d6df8f828aa6f43ad

**Comment by adam-fowler** [comment]
This has just been posted in the sourcekit-lsp documentation 
https://github.com/swiftlang/sourcekit-lsp/blob/main/Documentation/Using%20SourceKit-LSP%20with%20Embedded%20Projects.md
